rust-peer/src/main.rs

```diff
@@ -364,3 +428,91 @@ async fn read_or_create_identity(path: &Path) -> Result<identity::Keypair> {
 
     Ok(identity)
 }
+
+// Simple file exchange protocol
+
+#[derive(Debug, Clone)]
+struct FileExchangeProtocol();
+#[derive(Clone)]
+struct FileExchangeCodec();
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct FileRequest(String);
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct FileResponse(Vec<u8>);
+
+impl ProtocolName for FileExchangeProtocol {
+    fn protocol_name(&self) -> &[u8] {
+        "/universal-connectivity-file/1".as_bytes()
+    }
+}
+
+#[async_trait]
+impl request_response::Codec for FileExchangeCodec {
+    type Protocol = FileExchangeProtocol;
+    type Request = FileRequest;
+    type Response = FileResponse;
+
+    async fn read_request<T>(
+        &mut self,
+        _: &FileExchangeProtocol,
+        io: &mut T,
+    ) -> io::Result<Self::Request>
+    where
+        T: AsyncRead + Unpin + Send,
+    {
+        let vec = read_length_prefixed(io, 1_000_000).await?;
+
+        if vec.is_empty() {
+            return Err(io::ErrorKind::UnexpectedEof.into());
+        }
+
+        Ok(FileRequest(String::from_utf8(vec).unwrap()))
+    }
+
+    async fn read_response<T>(
+        &mut self,
+        _: &FileExchangeProtocol,
+        io: &mut T,
+    ) -> io::Result<Self::Response>
+    where
+        T: AsyncRead + Unpin + Send,
+    {
+        let vec = read_length_prefixed(io, 500_000_000).await?; // update transfer maximum
+
+        if vec.is_empty() {
+            return Err(io::ErrorKind::UnexpectedEof.into());
+        }
+
+        Ok(FileResponse(vec))
+    }
+
+    async fn write_request<T>(
+        &mut self,
+        _: &FileExchangeProtocol,
+        io: &mut T,
+        FileRequest(data): FileRequest,
+    ) -> io::Result<()>
+    where
+        T: AsyncWrite + Unpin + Send,
+    {
+        write_length_prefixed(io, data).await?;
+        io.close().await?;
```

Comment:
No need to close here, we do that already in `libp2p-request-response`.

Reply:
Thanks: d02649a. I was curious about this.